File: scripts/unconstrained_priced_interval.py
"""Linear exact fast path when the unconstrained priced optimum meets hard caps.

This solver considers all interval covers of supplied weighted/mandatory
sites without resource caps. A feasible unconstrained optimum is also a
hard-cap optimum. Callers must check both caps before using its witness.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping, Sequence


@dataclass(frozen=True)
class UnconstrainedCover:
    intervals: tuple[tuple[int, int], ...]
    mass: int
    units: int
    gets: int
    objective: int


@dataclass(frozen=True)
class _State:
    mass: int
    units: int
    gets: int
    prev_open: bool
    continued: bool


def _key(state: _State, unit_price: int, get_price: int) -> tuple[int, int, int, int]:
    return (state.mass - unit_price * state.units - get_price * state.gets,
            -state.units, -state.gets, state.mass)


def _prefer(left: _State | None, right: _State | None,
            unit_price: int, get_price: int) -> _State | None:
    if left is None:
        return right
    if right is None:
        return left
    return right if _key(right, unit_price, get_price) > _key(left, unit_price, get_price) else left


def unconstrained_priced_cover(
    weight_by_unit: Mapping[int, int], mandatory_units: Sequence[int], *,
    page_count: int, unit_price: int, get_price: int,
) -> UnconstrainedCover:
    """Optimize modeled mass minus nonnegative linear unit/GET prices."""
    mandatory = tuple(mandatory_units)
    if (type(page_count) is not int or page_count <= 0
            or type(unit_price) is not int or unit_price < 0
            or type(get_price) is not int or get_price < 0
            or len(set(mandatory)) != len(mandatory)
            or any(type(unit) is not int or not 0 <= unit < page_count
                   for unit in mandatory)
            or any(type(unit) is not int or not 0 <= unit < page_count
                   or type(weight) is not int or weight <= 0
                   for unit, weight in weight_by_unit.items())
            or sum(weight_by_unit.values()) >= 2**30):
        raise ValueError("unconstrained priced interval geometry differs")
    sites = sorted(set(mandatory) | set(weight_by_unit))
    if not sites:
        return UnconstrainedCover((), 0, 0, 0, 0)
    required = set(mandatory)
    closed: _State | None = _State(0, 0, 0, False, False)
    opened: _State | None = None
    history: list[tuple[int, bool | None, bool | None, bool | None]] = []
    previous: int | None = None
    for site in sites:
        weight = weight_by_unit.get(site, 0)
        gap = 1 if previous is None else site - previous
        if site in required:
            next_closed = None
        else:
            next_closed = _prefer(closed, opened, unit_price, get_price)
        closed_prev = None if next_closed is None else next_closed is opened
        starts = []
        if closed is not None:
            starts.append(_State(closed.mass + weight, closed.units + 1,
                                 closed.gets + 1, False, False))
        if opened is not None:
            starts.append(_State(opened.mass + weight, opened.units + 1,
                                 opened.gets + 1, True, False))
        new_open: _State | None = None
        for start in starts:
            new_open = _prefer(new_open, start, unit_price, get_price)
        continued = (None if opened is None else
                     _State(opened.mass + weight, opened.units + gap,
                            opened.gets, True, True))
        next_open = _prefer(new_open, continued, unit_price, get_price)
        history.append((site, closed_prev,
                        None if next_open is None else next_open.prev_open,
                        None if next_open is None else next_open.continued))
        closed, opened, previous = next_closed, next_open, site
    winner = _prefer(closed, opened, unit_price, get_price)
    assert winner is not None
    best_open = winner is opened
    pending_end: int | None = None
    intervals: list[tuple[int, int]] = []
    for site, closed_prev, open_prev, continued in reversed(history):
        if not best_open:
            if closed_prev is None:
                raise AssertionError("unconstrained closed backtrace differs")
            best_open = closed_prev
        else:
            if pending_end is None:
                pending_end = site
            if continued is None or open_prev is None:
                raise AssertionError("unconstrained open backtrace differs")
            if not continued:
                intervals.append((site, pending_end))
                pending_end = None
            best_open = open_prev
    if best_open or pending_end is not None:
        raise AssertionError("unconstrained backtrace origin differs")
    intervals.reverse()
    witness = tuple(intervals)
    mass = sum(weight for site, weight in weight_by_unit.items()
               if any(start <= site <= end for start, end in witness))
    units = sum(end - start + 1 for start, end in witness)
    gets = len(witness)
    objective = mass - unit_price * units - get_price * gets
    if (mass, units, gets, objective) != (
            winner.mass, winner.units, winner.gets,
            _key(winner, unit_price, get_price)[0]):
        raise AssertionError("unconstrained priced witness differs")
    if any(not any(start <= unit <= end for start, end in witness)
           for unit in mandatory):
        raise AssertionError("unconstrained mandatory witness differs")
    return UnconstrainedCover(witness, mass, units, gets, objective)

File: crates/borsuk/src/unconstrained_priced_interval.rs
//! Linear two-state priced interval cover, before hard resource admission.
//!
//! A caller may accept this optimum only when its witness meets both hard
//! caps. Otherwise it must use a capped solver. Weights are predicted mass,
//! not ground truth, and the unit/GET prices are supplied by the caller.

use std::collections::{BTreeMap, BTreeSet};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct UnconstrainedCover {
    pub intervals: Vec<(usize, usize)>,
    pub mass: i64,
    pub units: usize,
    pub gets: usize,
    pub objective: i64,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum CoverError {
    InvalidGeometry,
    ArithmeticOverflow,
    InvalidWitness,
}

#[derive(Clone, Copy)]
struct State {
    mass: i64,
    units: usize,
    gets: usize,
    prev_open: bool,
    continued: bool,
}

#[derive(Clone, Copy)]
struct Trace {
    site: usize,
    closed_prev_open: Option<bool>,
    open_prev_open: Option<bool>,
    open_continued: Option<bool>,
}

fn key(state: State, unit_price: i64, get_price: i64) -> (i64, i64, i64, i64) {
    let units = state.units as i64;
    let gets = state.gets as i64;
    (
        state.mass - unit_price * units - get_price * gets,
        -units,
        -gets,
        state.mass,
    )
}

fn prefer(
    left: Option<State>,
    right: Option<State>,
    unit_price: i64,
    get_price: i64,
) -> Option<State> {
    match (left, right) {
        (None, other) | (other, None) => other,
        (Some(a), Some(b)) => Some(
            if key(b, unit_price, get_price) > key(a, unit_price, get_price) {
                b
            } else {
                a
            },
        ),
    }
}

/// Maximize modeled mass minus priced physical units and GETs over all covers.
/// This is `O(weighted sites + mandatory sites)` time and trace space, aside
/// from ordered input construction. A feasible result is a hard-cap optimum.
pub fn unconstrained_priced_cover(
    weights: &BTreeMap<usize, i64>,
    mandatory: &[usize],
    page_count: usize,
    unit_price: i64,
    get_price: i64,
) -> Result<UnconstrainedCover, CoverError> {
    if page_count == 0
        || page_count > i64::MAX as usize
        || unit_price < 0
        || get_price < 0
        || unit_price
            .checked_add(get_price)
            .and_then(|sum| sum.checked_mul(page_count as i64))
            .is_none_or(|charge| charge > i64::MAX - (1 << 30))
    {
        return Err(CoverError::InvalidGeometry);
    }
    let required = mandatory.iter().copied().collect::<BTreeSet<_>>();
    if required.len() != mandatory.len()
        || required.iter().any(|site| *site >= page_count)
        || weights
            .iter()
            .any(|(&site, &weight)| site >= page_count || weight <= 0)
    {
        return Err(CoverError::InvalidGeometry);
    }
    let mut total_mass = 0_i64;
    for &weight in weights.values() {
        total_mass = total_mass
            .checked_add(weight)
            .ok_or(CoverError::ArithmeticOverflow)?;
    }
    if total_mass >= 1 << 30 {
        return Err(CoverError::InvalidGeometry);
    }
    let sites = required
        .iter()
        .copied()
        .chain(weights.keys().copied())
        .collect::<BTreeSet<_>>();
    if sites.is_empty() {
        return Ok(UnconstrainedCover {
            intervals: Vec::new(),
            mass: 0,
            units: 0,
            gets: 0,
            objective: 0,
        });
    }
    let mut closed = Some(State {
        mass: 0,
        units: 0,
        gets: 0,
        prev_open: false,
        continued: false,
    });
    let mut opened: Option<State> = None;
    let mut history = Vec::with_capacity(sites.len());
    let mut previous = None;
    for site in sites {
        let weight = *weights.get(&site).unwrap_or(&0);
        let gap = previous.map_or(1, |prev| site - prev);
        let (next_closed, closed_prev_open) = if required.contains(&site) {
            (None, None)
        } else {
            let from_open = match (closed, opened) {
                (Some(a), Some(b)) => key(b, unit_price, get_price) > key(a, unit_price, get_price),
                (None, Some(_)) => true,
                _ => false,
            };
            (
                prefer(closed, opened, unit_price, get_price),
                Some(from_open),
            )
        };
        let start_closed = closed.map(|state| State {
            mass: state.mass + weight,
            units: state.units + 1,
            gets: state.gets + 1,
            prev_open: false,
            continued: false,
        });
        let start_open = opened.map(|state| State {
            mass: state.mass + weight,
            units: state.units + 1,
            gets: state.gets + 1,
            prev_open: true,
            continued: false,
        });
        let new_interval = prefer(start_closed, start_open, unit_price, get_price);
        let continue_open = opened.map(|state| State {
            mass: state.mass + weight,
            units: state.units + gap,
            gets: state.gets,
            prev_open: true,
            continued: true,
        });
        let next_open = prefer(new_interval, continue_open, unit_price, get_price);
        history.push(Trace {
            site,
            closed_prev_open,
            open_prev_open: next_open.map(|state| state.prev_open),
            open_continued: next_open.map(|state| state.continued),
        });
        closed = next_closed;
        opened = next_open;
        previous = Some(site);
    }
    let winner = prefer(closed, opened, unit_price, get_price).ok_or(CoverError::InvalidWitness)?;
    let mut best_open = match (closed, opened) {
        (Some(a), Some(b)) => key(b, unit_price, get_price) > key(a, unit_price, get_price),
        (None, Some(_)) => true,
        _ => false,
    };
    let mut pending_end = None;
    let mut intervals = Vec::new();
    for trace in history.into_iter().rev() {
        if best_open {
            if pending_end.is_none() {
                pending_end = Some(trace.site);
            }
            let continued = trace.open_continued.ok_or(CoverError::InvalidWitness)?;
            if !continued {
                intervals.push((
                    trace.site,
                    pending_end.take().ok_or(CoverError::InvalidWitness)?,
                ));
            }
            best_open = trace.open_prev_open.ok_or(CoverError::InvalidWitness)?;
        } else {
            best_open = trace.closed_prev_open.ok_or(CoverError::InvalidWitness)?;
        }
    }
    if best_open || pending_end.is_some() {
        return Err(CoverError::InvalidWitness);
    }
    intervals.reverse();
    let mut units = 0_usize;
    for &(start, end) in &intervals {
        units = units
            .checked_add(end - start + 1)
            .ok_or(CoverError::ArithmeticOverflow)?;
    }
    let mut mass = 0_i64;
    let mut interval = 0_usize;
    for (&site, &weight) in weights {
        while interval < intervals.len() && intervals[interval].1 < site {
            interval += 1;
        }
        if interval < intervals.len() && intervals[interval].0 <= site {
            mass += weight;
        }
    }
    let gets = intervals.len();
    let objective = mass - unit_price * units as i64 - get_price * gets as i64;
    if (mass, units, gets, objective)
        != (
            winner.mass,
            winner.units,
            winner.gets,
            key(winner, unit_price, get_price).0,
        )
        || required
            .iter()
            .any(|site| !intervals.iter().any(|(a, b)| a <= site && site <= b))
    {
        return Err(CoverError::InvalidWitness);
    }
    Ok(UnconstrainedCover {
        intervals,
        mass,
        units,
        gets,
        objective,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn exact_small_witness_and_mandatory_bridge() {
        let weights = BTreeMap::from([(0, 10), (2, 8)]);
        let result = unconstrained_priced_cover(&weights, &[0], 3, 1, 3).unwrap();
        assert_eq!(result.intervals, vec![(0, 2)]);
        assert_eq!((result.mass, result.units, result.gets), (18, 3, 1));
        let strict = unconstrained_priced_cover(&weights, &[0, 2], 3, 1, 20).unwrap();
        assert_eq!(strict.intervals, vec![(0, 2)]);
    }
}
